File: 2707. Extra Characters in a String.py
class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        @cache
        def get_min_extra(i: int):
            if i == len(s):
                return 0
            res = 1 + get_min_extra(i + 1)
            for word in dictionary:
                if word == s[i:i + len(word)]:
                    res = min(res, get_min_extra(i + len(word)))
            return res
        
        return get_min_extra(0)

File: 2707. Extra Characters in a String.java
class Solution: 
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:

        words = defaultdict(list)
        for word in dictionary: words[word[0]].append(word)

        @lru_cache(1000)  
        def dp(ptr:int)-> int:

            if ptr == len(s): return 0

            cands = (w for w in words[s[ptr]] if s[ptr:].startswith(w))
            return min([dp(ptr+len(w)) for w in cands]+[1+dp(ptr+1)])

        return dp(0)
